File: python_helpers/rename_photos.py
import os
import re
import argparse

def natural_sort_key(s):
    return [int(text) if text.isdigit() else text.lower() for text in re.split(r'(\d+)', s)]

def extract_number(filename):
    match = re.search(r'\d+', filename)
    if match:
        return int(match.group())
    return None

def perform_renaming(seniors_dir, mode):
    if not os.path.exists(seniors_dir):
        return f"Error: Seniors directory does not exist at {seniors_dir}"

    subdirs = [d for d in os.listdir(seniors_dir) if os.path.isdir(os.path.join(seniors_dir, d))]
    total_renamed = 0
    report = []
    
    for subdir in sorted(subdirs):
        senior_name_prefix = "".join(subdir.split())
        subdir_path = os.path.join(seniors_dir, subdir)
        
        files = [f for f in os.listdir(subdir_path) if os.path.isfile(os.path.join(subdir_path, f))]
        sorted_files = sorted(files, key=natural_sort_key)
        
        report.append(f"\nProcessing Folder: {subdir}")
        report.append("-" * 40)
        
        if mode == "sequential":
            # Step 1: Rename to temporary names to prevent any collision
            temp_renames = []
            for idx, file in enumerate(sorted_files):
                file_path = os.path.join(subdir_path, file)
                ext = os.path.splitext(file)[1]
                temp_name = f"__temp_{idx:03d}__" + ext
                temp_path = os.path.join(subdir_path, temp_name)
                
                try:
                    os.rename(file_path, temp_path)
                    temp_renames.append((temp_path, idx, ext))
                except Exception as e:
                    report.append(f"  Error during temp rename of {file}: {e}")
                    
            # Step 2: Rename from temporary names to final sequential names
            for temp_path, idx, ext in temp_renames:
                final_name = f"{senior_name_prefix}_{idx+1:02d}{ext}"
                final_path = os.path.join(subdir_path, final_name)
                
                try:
                    os.rename(temp_path, final_path)
                    report.append(f"  {sorted_files[idx]} -> {final_name}")
                    total_renamed += 1
                except Exception as e:
                    report.append(f"  Error during final rename to {final_name}: {e}")
        else:
            # gap_fill mode: Keep existing numbers and fill remaining slots
            # Find which files have numbers <= 20
            numbered_files = []
            other_files = []
            for f in sorted_files:
                num = extract_number(f)
                if num is not None and num <= 20:
                    numbered_files.append((f, num))
                else:
                    other_files.append(f)
                    
            # Map of number -> original filename
            mapping = {}
            used_numbers = set()
            for f, num in numbered_files:
                mapping[num] = f
                used_numbers.add(num)
                
            # Fill gaps (1 to 20) with other files
            gaps = [n for n in range(1, 21) if n not in used_numbers]
            for idx, f in enumerate(other_files):
                if idx < len(gaps):
                    mapping[gaps[idx]] = f
                    
            # Perform rename via temp step to prevent collisions
            temp_renames = []
            for num in sorted(mapping.keys()):
                file = mapping[num]
                file_path = os.path.join(subdir_path, file)
                ext = os.path.splitext(file)[1]
                temp_name = f"__temp_{num:03d}__" + ext
                temp_path = os.path.join(subdir_path, temp_name)
                
                try:
                    os.rename(file_path, temp_path)
                    temp_renames.append((temp_path, num, ext, file))
                except Exception as e:
                    report.append(f"  Error during temp rename of {file}: {e}")
                    
            for temp_path, num, ext, orig_file in temp_renames:
                final_name = f"{senior_name_prefix}_{num:02d}{ext}"
                final_path = os.path.join(subdir_path, final_name)
                
                try:
                    os.rename(temp_path, final_path)
                    report.append(f"  {orig_file} -> {final_name}")
                    total_renamed += 1
                except Exception as e:
                    report.append(f"  Error during final rename to {final_name}: {e}")
                    
    report.append("\n" + "="*50)
    report.append("Renaming Complete!")
    report.append(f"Total files successfully numbered: {total_renamed}")
    report.append("="*50)
    return "\n".join(report)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Rename senior photos")
    parser.add_argument("--dir", required=True, help="Path to Seniors folder")
    parser.add_argument("--mode", default="sequential", choices=["sequential", "gap_fill"], help="Renaming mode")
    args = parser.parse_args()
    
    print(perform_renaming(args.dir, args.mode))

File: helpers/rename_photos.js
// helpers/rename_photos.js
import fs from 'node:fs/promises';
import path from 'node:path';
import { existsSync } from 'node:fs';

function naturalSortKey(s) {
  return s.split(/(\d+)/).map(text => {
    const num = parseInt(text, 10);
    return isNaN(num) ? text.toLowerCase() : num;
  });
}

function naturalCompare(a, b) {
  const keyA = naturalSortKey(a);
  const keyB = naturalSortKey(b);
  for (let i = 0; i < Math.max(keyA.length, keyB.length); i++) {
    if (keyA[i] === undefined) return -1;
    if (keyB[i] === undefined) return 1;
    if (keyA[i] !== keyB[i]) {
      if (typeof keyA[i] === 'number' && typeof keyB[i] === 'number') {
        return keyA[i] - keyB[i];
      }
      return String(keyA[i]).localeCompare(String(keyB[i]));
    }
  }
  return 0;
}

function extractNumber(filename) {
  const match = filename.match(/\d+/);
  return match ? parseInt(match[0], 10) : null;
}

export async function performRenaming(seniorsDir, mode) {
  if (!existsSync(seniorsDir)) {
    return `Error: Seniors directory does not exist at ${seniorsDir}`;
  }

  const entries = await fs.readdir(seniorsDir, { withFileTypes: true });
  const subdirs = entries
    .filter(e => e.isDirectory())
    .map(e => e.name)
    .sort(naturalCompare);

  let totalRenamed = 0;
  const report = [];

  for (const subdir of subdirs) {
    const seniorNamePrefix = subdir.replace(/\s+/g, '');
    const subdirPath = path.join(seniorsDir, subdir);

    const filesInDir = await fs.readdir(subdirPath, { withFileTypes: true });
    const files = filesInDir
      .filter(e => e.isFile())
      .map(e => e.name);
    
    const sortedFiles = files.sort(naturalCompare);

    report.push(`\nProcessing Folder: ${subdir}`);
    report.push("-".repeat(40));

    if (mode === "sequential") {
      // Step 1: Rename to temporary names to prevent any collision
      const tempRenames = [];
      for (let idx = 0; idx < sortedFiles.length; idx++) {
        const file = sortedFiles[idx];
        const filePath = path.join(subdirPath, file);
        const ext = path.extname(file);
        const tempName = `__temp_${String(idx).padStart(3, '0')}__${ext}`;
        const tempPath = path.join(subdirPath, tempName);

        try {
          await fs.rename(filePath, tempPath);
          tempRenames.push({ tempPath, idx, ext });
        } catch (e) {
          report.push(`  Error during temp rename of ${file}: ${e.message}`);
        }
      }

      // Step 2: Rename from temporary names to final sequential names
      for (const { tempPath, idx, ext } of tempRenames) {
        const finalName = `${seniorNamePrefix}_${String(idx + 1).padStart(2, '0')}${ext}`;
        const finalPath = path.join(subdirPath, finalName);

        try {
          await fs.rename(tempPath, finalPath);
          report.push(`  ${sortedFiles[idx]} -> ${finalName}`);
          totalRenamed++;
        } catch (e) {
          report.push(`  Error during final rename to ${finalName}: ${e.message}`);
        }
      }
    } else {
      // gap_fill mode: Keep existing numbers and fill remaining slots
      const numberedFiles = [];
      const otherFiles = [];

      for (const f of sortedFiles) {
        const num = extractNumber(f);
        if (num !== null && num <= 20) {
          numberedFiles.push({ file: f, num });
        } else {
          otherFiles.push(f);
        }
      }

      // Map of number -> original filename
      const mapping = new Map();
      const usedNumbers = new Set();

      for (const { file, num } of numberedFiles) {
        mapping.set(num, file);
        usedNumbers.add(num);
      }

      // Fill gaps (1 to 20) with other files
      const gaps = [];
      for (let n = 1; n <= 20; n++) {
        if (!usedNumbers.has(n)) {
          gaps.push(n);
        }
      }

      for (let idx = 0; idx < otherFiles.length; idx++) {
        const f = otherFiles[idx];
        if (idx < gaps.length) {
          mapping.set(gaps[idx], f);
        }
      }

      // Perform rename via temp step to prevent collisions
      const tempRenames = [];
      const sortedNums = Array.from(mapping.keys()).sort((a, b) => a - b);

      for (const num of sortedNums) {
        const file = mapping.get(num);
        const filePath = path.join(subdirPath, file);
        const ext = path.extname(file);
        const tempName = `__temp_${String(num).padStart(3, '0')}__${ext}`;
        const tempPath = path.join(subdirPath, tempName);

        try {
          await fs.rename(filePath, tempPath);
          tempRenames.push({ tempPath, num, ext, origFile: file });
        } catch (e) {
          report.push(`  Error during temp rename of ${file}: ${e.message}`);
        }
      }

      for (const { tempPath, num, ext, origFile } of tempRenames) {
        const finalName = `${seniorNamePrefix}_${String(num).padStart(2, '0')}${ext}`;
        const finalPath = path.join(subdirPath, finalName);

        try {
          await fs.rename(tempPath, finalPath);
          report.push(`  ${origFile} -> ${finalName}`);
          totalRenamed++;
        } catch (e) {
          report.push(`  Error during final rename to ${finalName}: ${e.message}`);
        }
      }
    }
  }

  report.push("\n" + "=".repeat(50));
  report.push("Renaming Complete!");
  report.push(`Total files successfully numbered: ${totalRenamed}`);
  report.push("=".repeat(50));
  return report.join("\n");
}

// Support command line usage
if (process.argv[1] && (process.argv[1].endsWith('rename_photos.js') || process.argv[1].endsWith('rename_photos.js"'))) {
  const args = process.argv.slice(2);
  let dir = '';
  let mode = 'sequential';

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--dir' && args[i + 1]) {
      dir = args[i + 1];
    } else if (args[i] === '--mode' && args[i + 1]) {
      mode = args[i + 1];
    }
  }

  if (!dir) {
    console.error("Error: --dir parameter is required.");
    process.exit(1);
  }

  performRenaming(dir, mode)
    .then(console.log)
    .catch(console.error);
}
